pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -86,13 +92,124 @@ type Interface interface {
+	volumeHost volume.VolumeHost,
+	migratedDrivers map[string](func() bool)) Interface {
 	return &nodeInfoManager{
-		nodeName:   nodeName,
-		volumeHost: volumeHost,
+		nodeName:        nodeName,
+		volumeHost:      volumeHost,
+		migratedDrivers: migratedDrivers,
 	}
 }
 
+// InitializeMigratedDrivers will create CSINodeInfo objects for all drivers that are to
+// be migrated. It will initialize most fields with dummy values. The CSIDriverInfo object
+// will be initialized with DriverName, ProxyInTreeVolumePlugin, and Available.
+func (nim *nodeInfoManager) InitializeMigratedDrivers() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		klog.Warningf("InitializeMigratedDrivers called but CSINodeInfo or CSIMigration is disabled")
+		return nil
+	}
+
+	var updateErrs []error
+
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	err := wait.ExponentialBackoff(updateBackoff, func() (bool, error) {
+		nodeInfo, err := csiKubeClient.CsiV1alpha1().CSINodeInfos().Get(string(nim.nodeName), metav1.GetOptions{})
+		if nodeInfo == nil || errors.IsNotFound(err) {
+			nodeInfo, err = nim.CreateCSINodeInfo()
+		}
+		if err != nil {
+			updateErrs = append(updateErrs, err)
+			return false, nil
+		}
+		err = nim.tryInitializeMigratedDrivers(nodeInfo)
+		if err != nil {
+			updateErrs = append(updateErrs, err)
+			return false, nil
+		}
+		return true, nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("error updating node: %v; caused by: %v", err, utilerrors.NewAggregate(updateErrs))
+	}
+	return nil
+}
+
+func (nim *nodeInfoManager) tryInitializeMigratedDrivers(nodeInfo *csiv1alpha1.CSINodeInfo) error {
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	existingDrivers := sets.String{}
+
+	// Do existing updates
+	// Clone driver list, omitting the driver that matches the given driverName
+	newDriverSpecs := []csiv1alpha1.CSIDriverInfoSpec{}
+	for _, driverInfoSpec := range nodeInfo.Spec.Drivers {
+		// No need to change any specs
+		newDriverSpecs = append(newDriverSpecs, driverInfoSpec)
+		existingDrivers.Insert(driverInfoSpec.Name)
+	}
+
+	newDriverStatuses := []csiv1alpha1.CSIDriverInfoStatus{}
+	for _, driverInfoStatus := range nodeInfo.Status.Drivers {
+		if _, ok := nim.migratedDrivers[driverInfoStatus.Name]; ok {
+			// If driver is known to migration logic and already exists in Spec
+			// just update volume plugin mechanism
+			driverInfoStatus.VolumePluginMechanism = nim.getVolumePluginMechanism(driverInfoStatus.Name)
+		}
+		newDriverStatuses = append(newDriverStatuses, driverInfoStatus)
+	}
+
+	for driver, _ := range nim.migratedDrivers {
+		if existingDrivers.Has(driver) {
+			// Skip adding new driver because already updated previously
+			continue
```

Comment:
1. Spec and status will eventually be modified in two seperate reconcile loops, spec will be reconciled from existance of driver sockets, and status will be reconciled based on availability of driver, though this has not been implemented yet. We will probably scope it for Q2.

2. I believe what you're referring to here is a reconciler style nodeinfomanager where the "authoritative database" would be the sockets in the plugin registration folder. Let me know if I misunderstood.

Kubelet should definitely not give up saving CSINodeInfo, I can fix that in this patch.

I believe there are no races between `tryInitializeMigratedDriver` and dynamic driver registration, since both of them retry on update conflict neither will stomp on the other changes. dynamic driver registration is considered authoritative, and `tryInitializeMigratedDriver` intention is to maintain the drivers registered dynamically as-is (maybe update VolumePluginMechanism) and only add new drivers for those that do not exist in CSINodeInfo thus far. I have test and verify this behavior in `nodeinfomanager_test.go`, if the code does not seem to be doing the right thing please point out specifics in the next round of review.

Thanks for the thorough review